keps/sig-storage/2639-secret-protection/README.md

```diff
@@ -0,0 +1,450 @@
+- Out-of-tree resources(`VolumeSnapshot`):
+  - The deletion is blocked by using newly introduced `snapshot.storage.kubernetes.io/secret-protection` finalizer,
+  - The `snapshot.storage.kubernetes.io/secret-protection` finalizer will be added on creation of the secret by using admission controller for `VolumeSnapshot`,
+  - The `snapshot.storage.kubernetes.io/secret-protection` finalizer will be deleted by newly introduced out-of-tree `secret-protection-volumesnapshot-controller` by checking whether the secret is in-use, on every change(Create/Update/Delete) events for secrets and related resources.
+
+Feature gate `SecretInUseProtection` is shared between in-tree and out-of-tree. Therefore, users won't be able to choose enabling secret portection for either in-tree or out-of-tree. On the other hand, users that doesn't install `VolumeSnapshot` feature won't be affected by secret protection for `VolumeSnapshot`.
+
+### Test Plan
+
+- For Alpha, unit tests and e2e tests verifying that a secret used by other resources is protected by this feature are added.
+- For Beta, scalability tests are added to exercise this feature.
+- For GA, the introduced e2e tests will be promoted to conformance.
+
+### Graduation Criteria
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- Allowing time for feedback
+
+#### Removing a Deprecated Flag
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+- Upgrade: Secret that doesn't have `kubernetes.io/secret-protection` finalizer and/or `snapshot.storage.kubernetes.io/secret-protection` finalizer will be added the finalizers on Update/Delete events, therefore no additional user operation will be needed.
+- Downgrade:
+  - Feature disabled case:
+    - If the `secret-protection-controller` exists and the feature is disabled, `kubernetes.io/secret-protection` finalizer will always be deleted, therefore no additional user operation will be needed,
```

Comment:
what will delete the finalizer on a downgrade to a version that lacks the feature?

Reply:
PVC protection controller always regards the PVC ["not being used"](https://github.com/kubernetes/kubernetes/blob/fbdc4f6acbc881eb26544d9d68ae520b7cbad3e7/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L222), if the feature gate is [disabled](https://github.com/kubernetes/kubernetes/blob/fbdc4f6acbc881eb26544d9d68ae520b7cbad3e7/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L254) . So, the controller delete the finalizer in this case (Controller is running even when feature gate is disabled, but not block the deletion). No controller case is described separately in the proposal for this reason.

Reply:
If the feature gate is enabled in K8S v1.22, but then it is downgraded to K8S v1.21, there could be secrets with protection finalizers?  How will those secrets be deleted?

Reply:
"Downgraded to no controller case" meant to cover this case. If it is confusing, could you advise a good wording for it? Changing former one to "Feature disabled with feature gate case"?